File: Day1/day1.cpp
#include <iostream>
#include <fstream>
#include <ostream>
#include <unordered_map>
#include <vector>
#include <algorithm>

int part1(std::vector<unsigned int> list1, std::vector<unsigned int> list2){
    std::sort(list1.begin(), list1.end());
    std::sort(list2.begin(), list2.end());
    
    unsigned int distance = 0;
    for (unsigned int i = 0; i < list1.size(); i++){
        distance += std::abs(static_cast<int>(list1[i]) - static_cast<int>(list2[i]));
    }
    return distance;
}

int part2(const std::vector<unsigned int>& list1, const std::vector<unsigned int>&list2){
    std::unordered_map<unsigned int, unsigned int> counter;
    for (unsigned int i = 0; i < list2.size(); i++){
        counter[list2[i]]++;
    }
    
    unsigned int similarity = 0;
    for (unsigned int i = 0; i < list1.size(); i++){
        similarity += list1[i] * counter[list1[i]];
    }
    return similarity;
}

int main(int argc, char *argv[]){
    std::vector<unsigned int> list1, list2;
    
    std::string sep = "/";
    std::string path(argv[0]);
    path = path.substr(0, path.rfind(sep));
    std::cout << path << std::endl;
    std::ifstream input;
    input.open(path + sep + "input.txt");
    
    unsigned int a, b;
    while (input >> a >> b){
        list1.push_back(a);
        list2.push_back(b);
    }
    input.close();

    std::cout << part1(list1, list2) << std::endl;
    std::cout << part2(list1, list2) << std::endl;
    
    return 0;
}

File: day1/day1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"slices"
)

func part1(array1 []int, array2 []int)(distance int){
	slices.Sort(array1)
	slices.Sort(array2)

	for i, _ := range array1 {
		dist := array2[i] - array1[i]
		// Amazingly, it seems there is no built-in abs function for ints...
		if dist > 0 { 
			distance += dist
		} else {
			distance += -dist
		}
	}
	return distance
}

func part2(array1 []int, array2 []int) (similarity int) {
	for _, v := range(array1) {
		nb_occurences := 0
		// ... and no built-in way to count the number of a specific element in a slice
		for _, w := range(array2){
			if v == w {
				nb_occurences += 1
			}
		}
		similarity += v * nb_occurences
	}
	return similarity
}

func main(){
	inFile, err := os.Open("day1.txt")
	if err != nil {
		panic(err)
	}
	defer inFile.Close()
	scanner := bufio.NewScanner(inFile)

	array1 := make([]int, 0)
	array2 := make([]int, 0)

	for scanner.Scan() {
		line := scanner.Text()
		nb, err := strconv.Atoi(line[:5])
		if err == nil{
			array1 = append(array1, nb)
		}
		
		nb, err = strconv.Atoi(line[len(line)-5:])
		if err == nil {
			array2 = append(array2, nb)
		}
	}

	
	fmt.Println(part1(array1[:],array2[:]))
	fmt.Println(part2(array1,array2))
}